Implement the power operation #3

MohamedAbdallah17 opened the pull request:
Resolves #3.

MohamedAbdallah17 committed:
Implement the power operation #3

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,57 @@ max:
+    la REG_PRINT_STRING_ARG,zero_exponent_message
+    jal printString
+    pop_ra_and_return
+  zero_exponent:
```
You probably don't need a special branch for this. See next comment.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,57 @@ max:
+  la REG_PRINT_STRING_ARG, result_message
   jal printString
+
+  # Power
+  add $t0, $zero, $zero 	#intialize $t0=0 , $t0 is used to record how many times we do multiplication operation  
+  add $t3, $t3, 1		#set intial value of $t3 = 1
+  power_loop_start:
+    beq $t0, $t2, power_loop_end 
+    mul $t3, $t3, $t1		#multiplication of $t1 and $t4 into $t4  
+    addi $t0,$t0,1		#update the value of $t0
+    j power_loop_start
+  power_loop_end:
```
You can start from the value of `b` and subtract one every loop, that was, if `b` is zero, the loop will not be executed and the output will be 1. This is similar to how we do the factorial above.

MohamedAbdallah17 replied:
when we say `mul $s3 ,$s1, s1` 
let's say we had a value saved in `$s3` ...is the result of multiplying `$s1` by `$s1` loaded into `$s3` neglecting the old value we had in `$s3` or adding it to old value ?.

MohamedAbdallah17 replied:
`# Power`
`  li $s3, $s1`
`  Power_loop_start:`
` li $t1, $zero`
 ` beqz $s2, Power_loop_end`
  `  mul $t1, $s3, $s1`
  `  add $s3, $s3, $t1`
  `  sub $s2, $s2, 1`
  `  b  Power_loop_start`
`  Power_loop_end:`
is this what you meant or not ? `$s1` is `a` and `$s2` is `b` and `$s3` is the result that would be printed after that 
How could i consider the zero case more efficiently not like the way i implemented it before ,, and i'll finish the case of **if b is negative value** and provide it for discussing

OmarEmaraDev replied:
@MohamedAbdallah17 You can incorporate the zero case by initializing `$s3` to 1 instead. The following C code demonstrates the solution:
```c
int power(int a, int b) {
    int result = 1;
    while (1) {
        if (b == 0) {break;}
        result *= a;
        b--;
    }
    return result;
}
```
A for loop can be used here, but I used a while loop for easier translation to assembly.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,57 @@ max:
+
+  # Read a.
+  jal readInteger
+  move $t1, REG_READ_INTEGER_RET
```
Don't store values in temporary registers if you will be calling functions after. Those registers are not callee-saved, meaning their values may change across function calls. Use saved registers instead.

MohamedAbdallah17 replied:
You mean that i would use **$s0 - $s7**

OmarEmaraDev replied:
@MohamedAbdallah17 Yes.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,57 @@ max:
+
+  # Read b.
+  jal readInteger
+  move $t2, REG_READ_INTEGER_RET
```
Same as above. Also, this value maybe negative, which is a case you don't consider.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,57 @@ max:
   jal printString
+
+  # Power
+  add $t0, $zero, $zero 	#intialize $t0=0 , $t0 is used to record how many times we do multiplication operation  
```
This is equivalent to `move $t0, $zero`.
I would prefer if the comments where in their separate lines, not after the lines, like in the rest of the code.

MohamedAbdallah17 committed:
Implement the power operation #3 

Is there any errors to work with this modifications? I think there is no any conflicts in the loops or results I traced them ;if I'm wrong inform me. 
please, I need to know when I write mul $s3, $s1, $s2  and $s3 had an initial value then $s3 will be updated with the new value neglecting old one  or adding it, multiplying it.. 
Thank you for your patience and help.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,64 @@ max:
 
 .text
-power:
-  push_ra
```
`push_ra` is necessary here. Don't remove it.

MohamedAbdallah17 replied:
sorry I don't mean to remove it

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,64 @@ max:
+  jal readInteger
+  move $s2, REG_READ_INTEGER_RET
+
+  # special case if b equals a negative value.3
+  move $s4, $s1
+  Special_power_loop_start:
+  bltz $s2, Special_power_loop_end
+    mul $t1, $s3, $s1
+    add $s4, $s4, $t1
+    add $s2, $s2, 1
+    b Special_power_loop_start
+  Special_power_loop_end:
```
Negative input handing is much simpler than this.
1. If b is negative and a is zero, print an error message and return.
2. Store the sign bit of b.
3. Take the absolute value of the b.
4. Compute the power.
5. If the sign bit is set, compute the reciprocal of the power.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,64 @@ max:
+  
+  # Print the result message.
+  li $t1, 1
+  div $s5, $t1, $s4
```
`div` can't be used here. You need to move the value to Co-Processor 1, compute the reciprocal there using the `RECIP.S` instruction and print the float.

OmarEmaraDev commented on mainMARS.s:
```diff
@@ -217,12 +217,64 @@ max:
+    b  Power_loop_start
+  Power_loop_end:
+  # Print the result message.
+  la REG_PRINT_STRING_ARG, zero_exponent_message
```
Why are you printing this here? It seems like you are printing multiple results here.

MohamedAbdallah17 committed:
implement the power operation

First, I'm sorry for being late. this is my last implemented one. 
Now ,I  forgot  some things and i need some clarifying  again in the special case "if the power is b=negative " as my  implementation isn't  efficient  or  runnig in the best right way.

OmarEmaraDev reviewed (approved):
Negative input handing will be refactored in a follow up commit.

Thanks!
